File: ScriptureAloneCore/Sources/ScriptureAloneCore/Basemap.swift
import Foundation

/// The offline base map (Basemap.bin, built by Tools/build_context.py from Natural Earth):
/// land and lake polygons and river lines, each at a coarse and a fine level of detail.
public struct Basemap: Sendable {
    public enum Kind: UInt8, Sendable { case land = 0, lake = 1, river = 2 }
    public enum Detail: UInt8, Sendable { case coarse = 0, fine = 1 }

    public struct Ring: Sendable {
        /// Natural Earth scale rank: lower is more prominent.
        public let rank: Int
        /// Longitude, latitude pairs (x, y).
        public let points: [SIMD2<Double>]
    }

    public struct Layer: Sendable {
        public let kind: Kind
        public let detail: Detail
        public let rings: [Ring]
    }

    public enum DecodeError: Error { case badHeader, truncated }

    /// Longitude/latitude box the coordinates were quantized over.
    public let minLongitude: Double
    public let minLatitude: Double
    public let maxLongitude: Double
    public let maxLatitude: Double
    public let layers: [Layer]

    public func layer(_ kind: Kind, _ detail: Detail) -> Layer? {
        layers.first { $0.kind == kind && $0.detail == detail }
    }

    public init(data: Data) throws {
        var reader = Reader(data: data)
        guard try reader.bytes(4) == Array("SABM".utf8) else { throw DecodeError.badHeader }
        let version: UInt16 = try reader.read()
        guard version == 1 else { throw DecodeError.badHeader }
        let layerCount: UInt16 = try reader.read()
        let x0 = Double(try reader.read() as Float32)
        let y0 = Double(try reader.read() as Float32)
        let x1 = Double(try reader.read() as Float32)
        let y1 = Double(try reader.read() as Float32)
        minLongitude = x0
        minLatitude = y0
        maxLongitude = x1
        maxLatitude = y1
        let sx = (x1 - x0) / 65535
        let sy = (y1 - y0) / 65535
        var layers: [Layer] = []
        for _ in 0..<layerCount {
            let kind: UInt8 = try reader.read()
            let detail: UInt8 = try reader.read()
            let _: UInt16 = try reader.read()
            let ringCount: UInt32 = try reader.read()
            var rings: [Ring] = []
            rings.reserveCapacity(Int(ringCount))
            for _ in 0..<ringCount {
                let count: UInt32 = try reader.read()
                let rank: UInt8 = try reader.read()
                let _: UInt8 = try reader.read()
                let _: UInt16 = try reader.read()
                var points: [SIMD2<Double>] = []
                points.reserveCapacity(Int(count))
                for _ in 0..<count {
                    let x: UInt16 = try reader.read()
                    let y: UInt16 = try reader.read()
                    points.append(SIMD2(x0 + Double(x) * sx, y0 + Double(y) * sy))
                }
                rings.append(Ring(rank: Int(rank), points: points))
            }
            if let kind = Kind(rawValue: kind), let detail = Detail(rawValue: detail) {
                layers.append(Layer(kind: kind, detail: detail, rings: rings))
            }
        }
        self.layers = layers
    }

    private struct Reader {
        let data: Data
        var offset = 0

        mutating func read<T: FixedWidthInteger>() throws -> T {
            let size = MemoryLayout<T>.size
            guard offset + size <= data.count else { throw DecodeError.truncated }
            let value = data.withUnsafeBytes { $0.loadUnaligned(fromByteOffset: offset, as: T.self) }
            offset += size
            return T(littleEndian: value)
        }

        mutating func read() throws -> Float32 {
            Float32(bitPattern: try read() as UInt32)
        }

        mutating func bytes(_ count: Int) throws -> [UInt8] {
            guard offset + count <= data.count else { throw DecodeError.truncated }
            defer { offset += count }
            return Array(data[data.startIndex + offset ..< data.startIndex + offset + count])
        }
    }
}

File: android/app/src/main/java/com/blainemiller/scripturealone/data/context/Basemap.kt
package com.blainemiller.scripturealone.data.context

/**
 * The offline base map (`Basemap.bin`, built by `Tools/build_context.py` from Natural Earth): land and
 * lake polygons and river lines, each at a coarse and a fine level of detail. Ported from
 * `ScriptureAloneCore/Basemap.swift`; both apps read the same file.
 *
 * The format, all little-endian:
 *
 *     "SABM"  u16 version (1)  u16 layerCount  f32 minLon  f32 minLat  f32 maxLon  f32 maxLat
 *     per layer:  u8 kind  u8 detail  u16 (reserved)  u32 ringCount
 *       per ring:  u32 pointCount  u8 rank  u8 (reserved)  u16 (reserved)
 *         per point:  u16 x  u16 y   — quantized over the bounding box, 0…65535
 *
 * A layer whose kind or detail this reader doesn't know is read past and dropped, so a newer file with
 * an extra layer still opens.
 */
class Basemap(data: ByteArray) {

    enum class Kind(val raw: Int) { LAND(0), LAKE(1), RIVER(2) }
    enum class Detail(val raw: Int) { COARSE(0), FINE(1) }

    /** A longitude (x), latitude (y) pair — Swift's `SIMD2<Double>`. */
    data class Point(val x: Double, val y: Double)

    class Ring(
        /** Natural Earth scale rank: lower is more prominent. */
        val rank: Int,
        /** Longitude, latitude pairs. */
        val points: List<Point>,
    )

    class Layer(val kind: Kind, val detail: Detail, val rings: List<Ring>)

    sealed class DecodeError(message: String) : Exception(message) {
        class BadHeader : DecodeError("not a Scripture Alone base map")
        class Truncated : DecodeError("the base map is truncated")
    }

    /** Longitude/latitude box the coordinates were quantized over. */
    val minLongitude: Double
    val minLatitude: Double
    val maxLongitude: Double
    val maxLatitude: Double
    val layers: List<Layer>

    fun layer(kind: Kind, detail: Detail): Layer? = layers.firstOrNull { it.kind == kind && it.detail == detail }

    init {
        val reader = Reader(data)
        if (!reader.bytes(4).contentEquals("SABM".toByteArray(Charsets.US_ASCII))) throw DecodeError.BadHeader()
        val version = reader.u16()
        if (version != 1) throw DecodeError.BadHeader()
        val layerCount = reader.u16()
        val x0 = reader.f32().toDouble()
        val y0 = reader.f32().toDouble()
        val x1 = reader.f32().toDouble()
        val y1 = reader.f32().toDouble()
        minLongitude = x0
        minLatitude = y0
        maxLongitude = x1
        maxLatitude = y1
        val sx = (x1 - x0) / 65535
        val sy = (y1 - y0) / 65535
        val layers = ArrayList<Layer>()
        repeat(layerCount) {
            val kind = reader.u8()
            val detail = reader.u8()
            reader.u16()
            val ringCount = reader.u32()
            // Counts come from the file, so capacity is capped by what the bytes could possibly hold
            // rather than trusted: a corrupt count then fails as Truncated instead of as an
            // allocation failure. (Swift reserves the full count.)
            val rings = ArrayList<Ring>(minOf(ringCount, reader.remaining / 8L).toInt())
            for (r in 0 until ringCount) {
                val count = reader.u32()
                val rank = reader.u8()
                reader.u8()
                reader.u16()
                val points = ArrayList<Point>(minOf(count, reader.remaining / 4L).toInt())
                for (p in 0 until count) {
                    val x = reader.u16()
                    val y = reader.u16()
                    points += Point(x0 + x * sx, y0 + y * sy)
                }
                rings += Ring(rank, points)
            }
            val k = Kind.entries.firstOrNull { it.raw == kind }
            val d = Detail.entries.firstOrNull { it.raw == detail }
            if (k != null && d != null) layers += Layer(k, d, rings)
        }
        this.layers = layers
    }

    /** Little-endian reads that throw [DecodeError.Truncated] rather than run off the end. */
    private class Reader(private val data: ByteArray) {
        private var offset = 0

        val remaining: Long get() = (data.size - offset).toLong()

        private fun take(size: Int): Int {
            if (offset + size > data.size) throw DecodeError.Truncated()
            val at = offset
            offset += size
            return at
        }

        fun u8(): Int = data[take(1)].toInt() and 0xFF

        fun u16(): Int {
            val at = take(2)
            return (data[at].toInt() and 0xFF) or ((data[at + 1].toInt() and 0xFF) shl 8)
        }

        fun u32(): Long {
            val at = take(4)
            return (data[at].toLong() and 0xFF) or ((data[at + 1].toLong() and 0xFF) shl 8) or
                ((data[at + 2].toLong() and 0xFF) shl 16) or ((data[at + 3].toLong() and 0xFF) shl 24)
        }

        fun f32(): Float = Float.fromBits(u32().toInt())

        fun bytes(count: Int): ByteArray {
            val at = take(count)
            return data.copyOfRange(at, at + count)
        }
    }
}
